src/Module.Server/Modes/Battle/FlagSystems/CrpgBattleFlagSystem.cs

```diff
@@ -0,0 +1,93 @@
+﻿using NetworkMessages.FromServer;
+using TaleWorlds.Core;
+using TaleWorlds.Library;
+using TaleWorlds.MountAndBlade;
+using TaleWorlds.MountAndBlade.Objects;
+using Timer = TaleWorlds.Core.Timer;
+
+namespace Crpg.Module.Modes.Battle.FlagSystems;
+internal class CrpgBattleFlagSystem : AbstractFlagSystem
+{
+    private int initiallySpawnedPlayerCount;
+
+    public CrpgBattleFlagSystem(Mission mission, MultiplayerGameNotificationsComponent notificationsComponent, CrpgBattleClient battleClient)
+        : base(mission, notificationsComponent, battleClient)
+    {
+    }
+
+    public override void CheckForManipulationOfFlags()
+    {
+        if (WereFlagsManipulated())
+        {
+            return;
+        }
+
+        Timer checkFlagRemovalTimer = GetCheckFlagRemovalTimer(Mission.CurrentTime, GetBattleClient().FlagManipulationTime);
+        if (!checkFlagRemovalTimer.Check(Mission.CurrentTime))
+        {
+            return;
+        }
+
+        var randomFlag = GetRandomFlag();
+        SpawnFlag(randomFlag);
+
+        SetWereFlagsManipulated(true);
+
+        GameNetwork.BeginBroadcastModuleEvent();
+        GameNetwork.WriteMessage(new CrpgBattleSpawnFlagMessage
+        {
+            FlagChar = randomFlag.FlagChar,
+        });
+        GameNetwork.EndBroadcastModuleEvent(GameNetwork.EventBroadcastFlags.None);
+
+        GetBattleClient().ChangeNumberOfFlags();
+        Debug.Print("Random Flag has been spawned");
+    }
+
+    public void IncrementInitiallySpawnedPlayerCount()
+    {
+        initiallySpawnedPlayerCount++;
+    }
+
+    public void CheckForDeadPlayerFlagSpawnThreshold()
+    {
+        float currentSpawnedPlayerCount = GetCurrentSpawnedPlayerCount();
+        if (currentSpawnedPlayerCount / initiallySpawnedPlayerCount <= 0.4f)
+        {
+            GetCheckFlagRemovalTimer(Mission.CurrentTime, GetBattleClient().FlagManipulationTime).Reset(Mission.CurrentTime, 0);
+        }
+    }
```

Comment:
public void CheckForDeadPlayerFlagSpawnThreshold()
    {
        int attackerCount = Mission.AttackerTeam.ActiveAgents.Count;
        int defenderCount = Mission.DefenderTeam.ActiveAgents.Count;
        // Todo : To move to server configuration
        float overpowerTheshold = 2f;
        if (attackerCount == 1 || defenderCount == 1)
        {
            // spawn flag
        }

        if (attackerCount > 7 || defenderCount > 7)
        {
            return;
        }

        if (Crpg.Module.Helpers.MathHelper.Within((float)attackerCount / (float)defenderCount, 1 / overpowerTheshold, overpowerTheshold))
        {
            return;
        }

        // spawn flag
    }